### PR TITLE
Player character list

### LINKED ISSUE #149: Memory: List of current party
In order to equip gear, we need to know the current party order. The equipment menu order changes depends on who is swapped in according to the Order menu.

Suggestion: an addition to `PlayerPartyManager` that exposes a `list[PlayerPartyCharacter]`.

### PR DESCRIPTION
**Pull request by Eein**
Adds an ordered player character list to player party manager

closes #149

based on a menu with the following:
```
______________________
|  char 0  |  char 3 |
|  char 1  |  char 4 |
|  char 2  |         |
```

the following:

image

will provide:

[<PlayerPartyCharacter.Valere: 'Valere'>, <PlayerPartyCharacter.Reshan: "Resh'an">, <PlayerPartyCharacter.Serai: 'Seraï'>, <PlayerPartyCharacter.Zale: 'Zale'>, <PlayerPartyCharacter.Bst: "B'st">]

### CODE REVIEW

File: memory/player_party_manager.py [hunk 1]
```diff
@@ -100,6 +148,9 @@ def _read_movement_state(self: Self) -> None:
                 case _:
                     self.movement_state = PlayerMovementState.NONE
 
+    def _read_players(self: Self) -> None:
```
Comment (orkaboy) [inline memory/player_party_manager.py:151]: This appears to be unused and not do anything.

File: memory/player_party_manager.py [hunk 2]
```diff
@@ -46,10 +52,50 @@ def update(self: Self) -> None:
...
+
             except Exception as _e:
                 # logger.debug(f"PlayerPartyManager Reloading {type(_e)}")
                 self.__init__()
 
+    def _read_current_party(self: Self) -> [PlayerPartyCharacter]:
+        try:
+            current_party_ptr = self.memory.follow_pointer(self.base, [0x98, 0x0])
+        except Exception:
+            self.players = []
```
Comment (orkaboy) [inline memory/player_party_manager.py:65]: There seems to be a mix and match between `self.players` and `self.current_party` here, with `self.current_party` seeming to be the "real" one. References to `self.players` should be removed.

Additionally, `self.current_party` should be assigned a default value and type in the `__init__`.

File: memory/player_party_manager.py [hunk 3]
```diff
@@ -15,6 +15,10 @@ class PlayerMovementState(Enum):
 
 
 class PlayerPartyManager:
+    NULL_POINTER = 0xFFFFFFFF
+    ITEM_OBJECT_OFFSET = 0x8
```
Comment (orkaboy) [inline memory/player_party_manager.py:19]: `ITEM_OBJECT_OFFSET` and `ITEM_INDEX_0_ADDRESS` are potentially badly named here. They do point to a list with items, yes, but it should be possible to see what kind of items (player characters in this case?) from the names, especially if we'll need to parse other lists in this class at some point.

File: memory/player_party_manager.py [hunk 4]
```diff
@@ -46,10 +52,50 @@ def update(self: Self) -> None:
...
+                ptr = self.memory.follow_pointer(items, [address, 0x0])
+                definition_id = self.memory.read_string(ptr + 0x14, 8)
+
+                # Definition IDS are stored as some goofy serialized utf encoded string
+                # We just do our best with the values that are provided to
+                # Determine the character we are looking at
+                players.append(PlayerPartyCharacter.parse_definition_id(definition_id))
+                address += self.ITEM_OBJECT_OFFSET
+            self.current_party = players
+        self.current_party = []
```
Comment (orkaboy) [inline memory/player_party_manager.py:97]: I don't see an early return or else in this function, so doesn't this undo all the work we just did?

File: memory/player_party_manager.py [hunk 5]
```diff
@@ -58,42 +59,36 @@ def update(self: Self) -> None:
...
-            return None
+            return
         # Item is an array of pointers of size 0x08
-        items = self.memory.follow_pointer(
+        # this follows playerPartyManager -> 0x98 (currentParty)
+        combat_players = self.memory.follow_pointer(
             current_party_ptr,
             [0x10, 0x0],
         )
         players = []
```
Comment (orkaboy) [inline memory/player_party_manager.py:78]: This can be moved to just before the for loop.

File: memory/player_party_manager.py [hunk 6]
```diff
@@ -58,42 +59,36 @@ def update(self: Self) -> None:
                 # logger.debug(f"PlayerPartyManager Reloading {type(_e)}")
                 self.__init__()
 
-    def _read_current_party(self: Self) -> [PlayerPartyCharacter]:
+    def _read_current_party(self: Self) -> None:
         try:
```
Comment (orkaboy) [inline memory/player_party_manager.py:63]: Potentially this try should be for the entire function? Can the call `combat_players = self.memory.follow_pointer(...` fail?

Comment (orkaboy) [reply memory/player_party_manager.py:63]: Also, completely missed the weird return value :P

Comment (Eein) [reply memory/player_party_manager.py:63]: All memory can but not likely if the main one fails. This one would be a nice crash to know about if it does

**Review by orkaboy — APPROVED** [review]

**Pull request merged**